### PR TITLE
Add several tests for sockets

### PR DESCRIPTION
**Pull request by LuKks**
For example, I set a new recv buffer size, so I get the recv buffer size and the size returned is double the one I set. Same happens with the send buffer size.

The tests are passing because at the moment of comparison I multiplied by two the new buffer size, but that's obviously not good.

Also, how do I know that the TTL change is actually working? Or is that enough?

### CODE REVIEW

**Comment by LuKks** [comment]
I thought GitHub tests were actually getting the correct size, so I pushed changes to remove the *2 trick, but it still fails so I saw wrong or not sure what's going on exactly.

**Comment by LuKks** [comment]
Why doing it this way breaks everything?

File: lib/socket.js [hunk 1]
```diff
@@ -138,7 +138,7 @@ module.exports = class UDXSocket extends events.EventEmitter {
     this.emit('listening')
   }
 
-  async close () {
+  close () {
```
Comment (mafintosh) [inline lib/socket.js:141]: Should be async so errors are wrapped

File: test/socket.js [hunk 2]
```diff
@@ -311,3 +312,284 @@ test('try send after close', async function (t) {
...
+  try {
+    a.setRecvBufferSize(NEW_BUFFER_SIZE)
+  } catch (error) {
+    t.is(error.message, 'Socket not active')
+  }
+
+  a.bind(0)
+  a.setRecvBufferSize(NEW_BUFFER_SIZE)
+
+  t.is(a.getRecvBufferSize(), NEW_BUFFER_SIZE)
```
Comment (mafintosh) [inline test/socket.js:534]: Api does not guarantee this, due to libuv not guaranteeing this. Test that it’s larger or equal or remove this test

File: test/socket.js [hunk 3]
```diff
@@ -311,3 +312,284 @@ test('try send after close', async function (t) {
...
+  const a = u.createSocket()
+
+  try {
+    a.getSendBufferSize()
+  } catch (error) {
+    t.is(error.message, 'Socket not active')
+  }
+
+  a.bind(0)
+  t.ok(a.getSendBufferSize() > 0)
```
Comment (mafintosh) [inline test/socket.js:552]: Same as above, re libuv

File: test/socket.js [hunk 4]
```diff
@@ -311,3 +312,284 @@ test('try send after close', async function (t) {
...
+  t.is(UDXSocket.isIPv6('127.0.0.1'), false)
+  t.is(UDXSocket.isIPv6('::1'), true)
+  t.is(UDXSocket.isIPv6('0.-1.0.0'), false)
+})
+
+test('UDXSocket - isIP', async function (t) {
+  t.is(UDXSocket.isIP('127.0.0.1'), 4)
+  t.is(UDXSocket.isIP('::1'), 6)
+  t.is(UDXSocket.isIP('0.-1.0.0'), 0)
+})
```
Comment (mafintosh) [inline test/socket.js:595]: None of these last three tests should be async

File: test/socket.js [hunk 5]
```diff
@@ -311,3 +312,284 @@ test('try send after close', async function (t) {
...
+  try {
+    a.setSendBufferSize(NEW_BUFFER_SIZE)
+  } catch (error) {
+    t.is(error.message, 'Socket not active')
+  }
+
+  a.bind(0)
+  a.setSendBufferSize(NEW_BUFFER_SIZE)
+
+  t.is(a.getSendBufferSize(), NEW_BUFFER_SIZE)
```
Comment (mafintosh) [inline test/socket.js:574]: Same as above

**Comment by LuKks** [comment]
For example, is it okay for `try send simple message` test to be an async function? Like it's considered async because there is something in the background (in this case waiting a message event, etc)? Or should I remove the async on those cases?

**Comment by mafintosh** [comment]
@LuKks good to go?

**Comment by LuKks** [comment]
@mafintosh yeap

File: test/socket.js [hunk 6]
```diff
@@ -311,3 +312,286 @@ test('try send after close', async function (t) {
...
+  })
+
+  b.bind(0)
+  a.trySend(Buffer.from('hello'), b.address().port)
+})
+
+test('get address without bind', async function (t) {
+  const u = new UDX()
+  const a = u.createSocket()
+  t.is(a.address(), null)
```
Comment (mafintosh) [inline test/socket.js:427]: close the socket here also after

File: test/socket.js [hunk 7]
```diff
@@ -311,3 +312,287 @@ test('try send after close', async function (t) {
...
+  a.trySend(Buffer.from('hello'), b.address().port)
+})
+
+test('get address without bind', async function (t) {
+  const u = new UDX()
+  const a = u.createSocket()
+
+  t.is(a.address(), null)
+
+  a.close()
```
Comment (LuKks) [inline test/socket.js:428]: Why adding `await` to `a.close()` ruins all the tests?

File: test/socket.js (hunk 6 above)
Comment (LuKks) [reply test/socket.js:431]: `bind()` sets `this._inited`, and the test doesn't bind, so the `close` event would never fire:
https://github.com/hyperswarm/libudx/blob/f01f56c236f3ec31ae4981c2e49d551171d1c650/lib/socket.js#L149
So if a socket didn't bind then I think there is no need to close

File: test/socket.js (hunk 7 above)
Comment (LuKks) [reply test/socket.js]: Quite literally, adding "await a.close()":
image

**Pull request merged**